packages/xod-project/src/extractBoundInputsToConstNodes.js

```diff
@@ -113,12 +113,15 @@ const getMapOfExtractablePinPaths = def(
   R.compose(
     R.map(R.reject(R.isNil)),
     R.map(
-      R.map(
-        ([pinType, pinValue]) =>
-          pinType === PIN_TYPE.PULSE
-            ? PULSE_CONST_NODETYPES[ensureLiteral(pinType, pinValue)]
-            : CONST_NODETYPES[pinType]
-      )
+      R.map(([pinType, pinValue]) => {
+        if (pinType === PIN_TYPE.PULSE) {
+          return R.compose(
+            foldMaybe(null, R.prop(R.__, PULSE_CONST_NODETYPES)),
+            ensureLiteral
+          )(pinType, pinValue);
+        }
+        return CONST_NODETYPES[pinType];
```

Comment:
Hard to understand.

FP vs plain old JS turbulence detected. Looking at the two top functions of the composition I see you’re trying to (a) convert pinType + pinValue to paths which can fail, then (b) reject failures. Nulls/maybes mixed which leads to hard readability. What if:

```js
R.compose(
  R.map(catMaybies)
  R.map(stuffToRejectNothings)
  R.map(R.map((...) => pinType === PIN_TYPE.PULSE
    ? maybeProp(PULSE_CONST_NODETYPES, ensureLiteral(...))
    : maybeProp(CONST_NODETYPES, pinType)))
```